File: sdk/typescript/src/policy-cache.ts
/**
 * Policy cache — fetches compiled policies from the SauronID core server
 * and keeps them hot in memory for sub-millisecond local evaluation.
 *
 * The cache fetches via `GET /v1/policy/:id` and stores the structured
 * AST. An optional background timer refreshes each policy on a fixed
 * interval; failed refreshes log a warning and keep the last good copy.
 *
 * @module
 */

/** Classification-tag based data scope. Mirrors the server `DataScope` AST node. */
export interface DataScopeShape {
    /** Tags the agent may operate on. Empty = no allowlist constraint. */
    allow: string[];
    /** Tags the agent must never touch. Takes precedence over `allow`. */
    deny: string[];
}

/** Rate limit (requests per minute). Mirrors server `RateLimit`. */
export interface RateLimitShape {
    /** Maximum requests per minute. */
    requests_per_minute: number;
}

/** Wall-clock window the agent may act within. Mirrors server `TimeWindow`. */
export interface TimeWindowShape {
    /** Window start, `HH:MM` 24-hour. */
    start: string;
    /** Window end, `HH:MM` 24-hour. */
    end: string;
    /** IANA timezone (e.g. `Europe/Paris`). */
    timezone: string;
}

/** M-of-N signature requirement clause. Mirrors server `SignatureRequirement`. */
export interface SignatureRequirementShape {
    /** Role name expected to sign (e.g. `human_approver`). */
    role: string;
    /** Minimum number of signatures of that role. */
    threshold: number;
}

/** Limits on sub-agent delegation. Mirrors server `DelegationLimits`. */
export interface DelegationLimitsShape {
    /** Maximum delegation depth (0 disables delegation). */
    max_depth: number;
    /** Allowed sub-agent identifiers. */
    allowed_subagents?: string[];
}

/** Structured binding section returned by `GET /v1/policy/:id`. */
export interface BindingShape {
    /** Tool allowlist. Absent = no constraint. */
    allowed_tools?: string[];
    /** Max cumulative spend in USD. Absent = no cap. */
    max_budget_usd?: number;
    /** Classification-based data scope. */
    data_scope?: DataScopeShape;
    /** Rate limit. */
    rate_limit?: RateLimitShape;
    /** Wall-clock window. */
    time_window?: TimeWindowShape;
    /** M-of-N signature requirements. */
    required_signatures?: SignatureRequirementShape[];
    /** Delegation limits. */
    delegation?: DelegationLimitsShape;
}

/**
 * Compiled policy as observed by the SDK. The server stores a richer
 * `CompiledPolicy` internally; we only need the fields that drive the
 * local invariant checks here.
 */
export interface CompiledPolicy {
    /** Server-assigned id (`pol_<32-hex>`). Echoed back from the caller. */
    policy_id: string;
    /** Agent identifier. */
    agent: string;
    /** DSL version. */
    version: string;
    /** Optional raw YAML, kept for diagnostics. Empty when not supplied. */
    raw_yaml: string;
    /** Names of the checks the policy compiled into. */
    checks: string[];
    /** Structured binding fields (allowed_tools, max_budget_usd, ...). */
    binding: BindingShape;
}

/** Shape of the JSON returned by `GET /v1/policy/:id` — the server `Policy` AST. */
interface ServerPolicyAst {
    version: string;
    agent: string;
    description?: string;
    binding?: BindingShape;
    invariants?: string[];
    metadata?: unknown;
}

/** Options for the {@link PolicyCache} constructor. */
export interface PolicyCacheOptions {
    /** Base URL of the core server (no trailing slash). */
    coreUrl: string;
    /** Admin auth bearer token, required by `/v1/policy/*` routes. */
    adminKey?: string;
    /** Background refresh interval in ms. Default 60_000. Set to 0 to disable. */
    refreshIntervalMs?: number;
    /** Override fetch (for tests / non-browser runtimes). */
    httpFetch?: typeof fetch;
    /**
     * Optional tenant id (Sprint 11). When set, every outbound request to
     * `/v1/policy/*` carries `x-sauron-tenant-id: <tenantId>`. When unset,
     * the request is treated as the `"default"` tenant on the server side,
     * preserving backwards compatibility with single-tenant deployments.
     */
    tenantId?: string;
}

/** In-memory policy cache with optional background refresh. */
export class PolicyCache {
    private readonly coreUrl: string;
    private readonly adminKey: string | undefined;
    private readonly refreshIntervalMs: number;
    private readonly httpFetch: typeof fetch;
    private readonly tenantId: string | undefined;

    private readonly entries: Map<string, CompiledPolicy> = new Map();
    private readonly timers: Map<string, ReturnType<typeof setInterval>> = new Map();

    constructor(opts: PolicyCacheOptions) {
        this.coreUrl = opts.coreUrl.replace(/\/+$/, "");
        this.adminKey = opts.adminKey;
        this.refreshIntervalMs = opts.refreshIntervalMs ?? 60_000;
        this.tenantId = opts.tenantId;
        const fallback = (globalThis as { fetch?: typeof fetch }).fetch;
        if (!opts.httpFetch && !fallback) {
            throw new Error(
                "PolicyCache: no fetch available — pass opts.httpFetch on Node < 18"
            );
        }
        this.httpFetch = (opts.httpFetch ?? fallback) as typeof fetch;
    }

    /**
     * Load a policy by id. Fetches from the server, caches the result,
     * and arms the background refresh timer. Returns the cached entry
     * on subsequent calls for the same id without a network roundtrip.
     */
    async load(policyId: string): Promise<CompiledPolicy> {
        const existing = this.entries.get(policyId);
        if (existing) return existing;
        const fresh = await this.fetchOne(policyId);
        this.entries.set(policyId, fresh);
        this.armRefresh(policyId);
        return fresh;
    }

    /** Synchronous cache read. Returns `undefined` on miss. */
    get(policyId: string): CompiledPolicy | undefined {
        return this.entries.get(policyId);
    }

    /**
     * Force a fresh fetch from the server. If the request fails, the
     * cached entry is preserved and a warning is logged.
     */
    async refresh(policyId: string): Promise<void> {
        try {
            const fresh = await this.fetchOne(policyId);
            this.entries.set(policyId, fresh);
        } catch (err) {
            // Keep last good copy; surface as warning.
            const msg = err instanceof Error ? err.message : String(err);
            // eslint-disable-next-line no-console
            console.warn(`[PolicyCache] refresh ${policyId} failed: ${msg}`);
        }
    }

    /** Stop every background refresh timer. Idempotent. Call before exit. */
    stop(): void {
        for (const t of this.timers.values()) clearInterval(t);
        this.timers.clear();
    }

    /** Internal: HTTP fetch + parse into {@link CompiledPolicy}. */
    private async fetchOne(policyId: string): Promise<CompiledPolicy> {
        const url = `${this.coreUrl}/v1/policy/${encodeURIComponent(policyId)}`;
        const headers: Record<string, string> = { accept: "application/json" };
        if (this.adminKey) headers.authorization = `Bearer ${this.adminKey}`;
        // Sprint 11: pass through the tenant header when the cache was
        // constructed with a tenant id. ADDITIVE only — single-tenant
        // deployments do not see this header and continue to land on the
        // default tenant on the server side.
        if (this.tenantId) headers["x-sauron-tenant-id"] = this.tenantId;
        const resp = await this.httpFetch(url, { method: "GET", headers });
        if (!resp.ok) {
            throw new Error(`GET ${url} → ${resp.status}`);
        }
        const ast = (await resp.json()) as ServerPolicyAst;
        const binding: BindingShape = ast.binding ?? {};
        // Derive `checks` from which binding fields are populated. The
        // server's compiler does the canonical mapping; this is a best
        // effort mirror for diagnostics and SHOULD match the server set.
        const checks: string[] = [];
        if (binding.allowed_tools) checks.push("allowlist");
        if (typeof binding.max_budget_usd === "number") checks.push("budget");
        if (binding.data_scope) checks.push("scope");
        if (binding.rate_limit) checks.push("rate_limit");
        if (binding.time_window) checks.push("time_window");
        if (binding.required_signatures && binding.required_signatures.length > 0) {
            checks.push("signatures");
        }
        if (binding.delegation) checks.push("delegation_depth");
        return {
            policy_id: policyId,
            agent: ast.agent,
            version: ast.version,
            raw_yaml: "",
            checks,
            binding,
        };
    }

    /** Internal: (re)arm the background refresh timer for a policy. */
    private armRefresh(policyId: string): void {
        const existing = this.timers.get(policyId);
        if (existing) clearInterval(existing);
        if (this.refreshIntervalMs <= 0) return;
        const t = setInterval(() => {
            void this.refresh(policyId);
        }, this.refreshIntervalMs);
        // Don't keep the Node event loop alive on the timer.
        if (typeof (t as { unref?: () => void }).unref === "function") {
            (t as { unref: () => void }).unref();
        }
        this.timers.set(policyId, t);
    }
}

File: sdk/go/sauronid/policy_cache.go
package sauronid

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"sync"
	"time"
)

// PolicyCacheOptions configures a PolicyCache.
type PolicyCacheOptions struct {
	// CoreURL is the base URL of the core server (no trailing slash).
	CoreURL string
	// AdminKey is the admin bearer token. Empty disables the header.
	AdminKey string
	// RefreshInterval is the background refresh period. Zero disables refresh.
	// Default 60s.
	RefreshInterval time.Duration
	// HTTPClient overrides the default http.Client (tests / custom transport).
	HTTPClient *http.Client
	// TenantID, when non-empty, attaches an X-Sauron-Tenant-Id header to every
	// outbound request. Single-tenant deployments leave this empty.
	TenantID string
}

// PolicyCache fetches compiled policies from the core server and keeps
// them hot in memory for sub-millisecond local evaluation.
//
// The cache is safe for concurrent use. Background refresh is armed on
// the first successful Load(); subsequent reads via Get() are read-only
// against the in-memory map.
type PolicyCache struct {
	coreURL         string
	adminKey        string
	refreshInterval time.Duration
	client          *http.Client
	tenantID        string

	mu      sync.RWMutex
	entries map[string]*CompiledPolicy
	timers  map[string]*time.Ticker
	stopChs map[string]chan struct{}
	stopped bool
}

// NewPolicyCache constructs a cache. Pass an empty struct for defaults.
func NewPolicyCache(opts PolicyCacheOptions) *PolicyCache {
	if opts.HTTPClient == nil {
		opts.HTTPClient = &http.Client{Timeout: 10 * time.Second}
	}
	if opts.RefreshInterval == 0 {
		opts.RefreshInterval = 60 * time.Second
	}
	return &PolicyCache{
		coreURL:         trimTrailingSlash(opts.CoreURL),
		adminKey:        opts.AdminKey,
		refreshInterval: opts.RefreshInterval,
		client:          opts.HTTPClient,
		tenantID:        opts.TenantID,
		entries:         map[string]*CompiledPolicy{},
		timers:          map[string]*time.Ticker{},
		stopChs:         map[string]chan struct{}{},
	}
}

// Load fetches the policy with the given id, caches it, and arms a
// background refresh timer. Subsequent calls for the same id return the
// cached entry without a network roundtrip.
func (c *PolicyCache) Load(ctx context.Context, policyID string) (*CompiledPolicy, error) {
	c.mu.RLock()
	existing, ok := c.entries[policyID]
	c.mu.RUnlock()
	if ok {
		return existing, nil
	}
	fresh, err := c.fetchOne(ctx, policyID)
	if err != nil {
		return nil, err
	}
	c.mu.Lock()
	c.entries[policyID] = fresh
	c.mu.Unlock()
	c.armRefresh(policyID)
	return fresh, nil
}

// Get returns the cached policy or nil on miss. It performs no I/O.
func (c *PolicyCache) Get(policyID string) *CompiledPolicy {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.entries[policyID]
}

// Refresh forces a fresh fetch. On failure the cached entry is preserved
// and a warning is logged via the standard logger.
func (c *PolicyCache) Refresh(ctx context.Context, policyID string) error {
	fresh, err := c.fetchOne(ctx, policyID)
	if err != nil {
		log.Printf("[PolicyCache] refresh %s failed: %v", policyID, err)
		return err
	}
	c.mu.Lock()
	c.entries[policyID] = fresh
	c.mu.Unlock()
	return nil
}

// Stop cancels every background refresh timer. Idempotent. Call before
// process exit.
func (c *PolicyCache) Stop() {
	c.mu.Lock()
	if c.stopped {
		c.mu.Unlock()
		return
	}
	c.stopped = true
	stopChs := c.stopChs
	timers := c.timers
	c.stopChs = map[string]chan struct{}{}
	c.timers = map[string]*time.Ticker{}
	c.mu.Unlock()
	for _, ch := range stopChs {
		close(ch)
	}
	for _, t := range timers {
		t.Stop()
	}
}

func (c *PolicyCache) fetchOne(ctx context.Context, policyID string) (*CompiledPolicy, error) {
	u := fmt.Sprintf("%s/v1/policy/%s", c.coreURL, url.PathEscape(policyID))
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/json")
	if c.adminKey != "" {
		req.Header.Set("Authorization", "Bearer "+c.adminKey)
	}
	if c.tenantID != "" {
		req.Header.Set("X-Sauron-Tenant-Id", c.tenantID)
	}
	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("GET %s -> %d: %s", u, resp.StatusCode, string(body))
	}
	var ast struct {
		Version     string  `json:"version"`
		Agent       string  `json:"agent"`
		Description string  `json:"description"`
		Binding     Binding `json:"binding"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&ast); err != nil {
		return nil, fmt.Errorf("decode policy %s: %w", policyID, err)
	}
	checks := deriveChecks(&ast.Binding)
	return &CompiledPolicy{
		PolicyID: policyID,
		Agent:    ast.Agent,
		Version:  ast.Version,
		Binding:  ast.Binding,
		Checks:   checks,
	}, nil
}

func (c *PolicyCache) armRefresh(policyID string) {
	if c.refreshInterval <= 0 {
		return
	}
	c.mu.Lock()
	if c.stopped {
		c.mu.Unlock()
		return
	}
	if old, ok := c.timers[policyID]; ok {
		old.Stop()
	}
	if oldCh, ok := c.stopChs[policyID]; ok {
		close(oldCh)
	}
	ticker := time.NewTicker(c.refreshInterval)
	stopCh := make(chan struct{})
	c.timers[policyID] = ticker
	c.stopChs[policyID] = stopCh
	c.mu.Unlock()
	go func() {
		for {
			select {
			case <-ticker.C:
				// Use a fresh context so cancelled parents do not kill refresh.
				ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
				_ = c.Refresh(ctx, policyID)
				cancel()
			case <-stopCh:
				return
			}
		}
	}()
}

// deriveChecks mirrors the TS / Python derivation so the diagnostic
// `Checks` slice is identical across SDKs.
func deriveChecks(b *Binding) []string {
	var checks []string
	if b.AllowedTools != nil {
		checks = append(checks, "allowlist")
	}
	if b.MaxBudgetUsd != nil {
		checks = append(checks, "budget")
	}
	if b.DataScope != nil {
		checks = append(checks, "scope")
	}
	if b.RateLimit != nil {
		checks = append(checks, "rate_limit")
	}
	if b.TimeWindow != nil {
		checks = append(checks, "time_window")
	}
	if len(b.RequiredSignatures) > 0 {
		checks = append(checks, "signatures")
	}
	if b.Delegation != nil {
		checks = append(checks, "delegation_depth")
	}
	return checks
}

func trimTrailingSlash(s string) string {
	for len(s) > 0 && s[len(s)-1] == '/' {
		s = s[:len(s)-1]
	}
	return s
}

// ErrPolicyNotFound is returned by Load when the server returns 404.
var ErrPolicyNotFound = errors.New("policy not found")
